src/app/labeling/qgsmaptoollabel.cpp

```diff
@@ -335,25 +335,84 @@ QString QgsMapToolLabel::currentLabelText( int trunc )
   return QString();
 }
 
-void QgsMapToolLabel::currentAlignment( QString &hali, QString &vali )
+QgsPalLayerSettings::QuadrantPosition QgsMapToolLabel::currentAlignment()
 {
-  hali = QStringLiteral( "Left" );
-  vali = QStringLiteral( "Bottom" );
+  QgsPalLayerSettings::QuadrantPosition quadrantPosition = QgsPalLayerSettings::QuadrantAboveRight;
 
   QgsVectorLayer *vlayer = mCurrentLabel.layer;
   if ( !vlayer )
   {
-    return;
+    return quadrantPosition;
   }
 
   QgsFeature f;
   if ( !currentFeature( f ) )
   {
-    return;
+    return quadrantPosition;
+  }
+
+  // quadrant offest defined via buttons
+  if ( mCurrentLabel.settings.placement == QgsPalLayerSettings::Placement::OverPoint )
+    quadrantPosition = mCurrentLabel.settings.quadOffset;
+
+  // data defined quadrant offset
+  if ( mCurrentLabel.settings.placement == QgsPalLayerSettings::Placement::AroundPoint ||
+       mCurrentLabel.settings.placement == QgsPalLayerSettings::Placement::OverPoint )
+  {
+    if ( mCurrentLabel.settings.dataDefinedProperties().isActive( QgsPalLayerSettings::OffsetQuad ) )
+    {
+      QVariant exprVal = evaluateDataDefinedProperty( QgsPalLayerSettings::OffsetQuad, mCurrentLabel.settings, f, quadrantPosition );
+      if ( !exprVal.isNull() )
+      {
+        bool ok;
+        int quadInt = exprVal.toInt( &ok );
+        if ( ok && 0 <= quadInt && quadInt <= 8 )
+        {
+          quadrantPosition = static_cast< QgsPalLayerSettings::QuadrantPosition >( quadInt );
+        }
+      }
+    }
+  }
+
+  // quadrant defined by DD alignment
+  if ( mCurrentLabel.settings.dataDefinedProperties().isActive( QgsPalLayerSettings::Hali ) ||
+       mCurrentLabel.settings.dataDefinedProperties().isActive( QgsPalLayerSettings::Vali ) )
+  {
+    QString hali = QStringLiteral( "Left" );
+    QString vali = QStringLiteral( "Bottom" );
+    hali = evaluateDataDefinedProperty( QgsPalLayerSettings::Hali, mCurrentLabel.settings, f, hali ).toString();
+    vali = evaluateDataDefinedProperty( QgsPalLayerSettings::Vali, mCurrentLabel.settings, f, vali ).toString();
+
+    if ( hali.compare( QLatin1String( "Left" ), Qt::CaseInsensitive ) == 0 )
+    {
+      if ( vali.compare( QLatin1String( "Top" ), Qt::CaseInsensitive ) == 0 || vali.compare( QLatin1String( "Cap" ), Qt::CaseInsensitive ) == 0 )
+        quadrantPosition = QgsPalLayerSettings::QuadrantBelowRight;
+      else if ( vali.compare( QLatin1String( "Half" ), Qt::CaseInsensitive ) == 0 )
```

Comment:
There's also the "Cap" mode -- this isn't the same as half, and unfortunately doesn't map to an existing quadrant value.

Potentially the quadrant enum could be extended to include the cap based placements....

Reply:
Oh I see, in the current implementation `Top` and `Cap` where treated in one block so I assumed they were the same.

Instead of extending the enum i would rename this function to `currentAlignmentOffset` and return directly `xdiff` and `ydiff`.

Reply:
@nyalldawson I was feeling wrong adding the Cap/Base alignments to an enum called "Quadrant". I added a new Alignment enum to QgsMapToolLabel for internal use. Are you ok with this?